### PR TITLE
feat: registry-driven layer defaults, param validation

### LINKED ISSUE #177: [Architecture Proposal] Phase 1 Foundation: Dynamic Data-Driven Layer Registry (Single Source of Truth)
## Overview

As outlined in the GSoC 2026 Phase 1 Expected Results, TensorMap needs to transition from a hardcoded layer system (requiring 6+ file edits) to a data-driven layer registry. 

To achieve this and easily scale to 15+ layer types (CNNs, RNNs, Regularization), I propose a **Single Source of Truth (SSOT)** architecture. By defining layers in a central JSON registry served by the backend, we can decouple the frontend UI from the backend generation logic.

## Current Bottleneck (The 6-File Problem)

Currently, adding a layer like `Dropout` requires:

1. Hardcoding the button in `Sidebar.jsx`
2. Creating a dedicated `DropoutNode.jsx` UI component
3. Registering the node in `Canvas.jsx`
4. Hardcoding Pydantic schemas in `deep_learning.py`
5. Updating the code generation engine
6. Updating the Keras compilation mapping

## Proposed Architecture: The Unified Registry

I propose creating a `layer_registry.json` on the backend and serving it via a new endpoint (`GET /api/v1/layers`). Both the FastAPI backend and React frontend will use this single contract.

### 1. The JSON Contract (Example)

```json
{
  "layers": {
    "dense": {
      "display_name": "Dense",
      "category": "core",
      "keras_mapping": "tf.keras.layers.Dense",
      "params": {
        "units": { "type": "number", "default": 32, "min": 1, "required": true },
        "activation": { "type": "select", "options": ["relu", "sigmoid", "softmax", "linear"], "default": "relu" }
      }
    },
    "dropout": {
      "display_name": "Dropout",
      "category": "regularization",
      "keras_mapping": "tf.keras.layers.Dropout",
      "params": {
        "rate": { "type": "number", "default": 0.5, "min": 0.0, "max": 1.0, "required": true }
      }
    }
  }
}
````

### 2. Frontend Implementation (Dynamic Rendering)

**Dynamic Sidebar:**
`Sidebar.jsx` fetches the registry on load and maps through layers to generate the draggable buttons, grouping them by category.

**Universal Node Component:**
Instead of a separate `.jsx` file for every layer, we build a single `GenericLayerNode.jsx`. When dropped on the canvas, it looks up its type in the registry and dynamically maps the `params` object to UI inputs (rendering a `<Select>` for `activation` and a number `<Input>` for `units`).

### 3. Backend Implementation (Dynamic Validation)

**Pydantic Dynamic Models:**
Instead of hardcoded models, we use `create_model` in Pydantic to dynamically generate validation schemas based on the JSON registry at server startup.

**Streamlined Code Gen:**
The generation engine loops through the ReactFlow payload, looks up the `keras_mapping` string, and dynamically unpacks the frontend payload into kwargs:

```python
f"{layer.keras_mapping}(**{layer.params})"
```
## Action Plan & Phased Rollout

If the maintainers are aligned with this architecture, I am ready to lead Phase 1 with the following PRs:

* **PR 1:** Define the `layer_registry.json` and build the `GET /api/v1/layers` endpoint.
* **PR 2:** Build the frontend `GenericLayerNode.jsx` to dynamically consume the API and replace the existing hardcoded Dense/Conv2D nodes.
* **PR 3:** Update backend code generation to use the registry dictionary.

@ivantha I'd love your thoughts on this SSOT approach. Does this align with your vision for the Phase 1 Foundation?

### PR DESCRIPTION
**Pull request by Incharajayaram**
## Description

Layer parameter defaults and required fields now driven by registry (not hardcoded)
Improved validation: Save is only enabled when all required params are filled and the graph is fully connected

Fixes #177 

## Type of Change

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Documentation update

Registry Structure
```
export const LAYERS = [
  {
    type: "customdense",
    label: "Dense",
    category: "Dense",
    params: {
      units: { type: "number", required: true, default: 64, min: 1 },
      activation: { type: "select", required: true, default: "relu", options: ["relu", "sigmoid", "tanh"] }
    }
  },
  // ...more layers
];
```
## How Has This Been Tested?

Describe the tests you ran to verify your changes.

- [x] Existing tests pass
- [ ] New tests added
- [x] Manual testing

Added all node types, verified default params
Connected nodes in order: Input → Conv2D → Flatten → Dense
Filled all required params, Save enabled
Negative cases: missing edge or param disables Save
Model save and reload works, graph/params persist
Sidebar drag-and-drop and registry defaults verified

## Checklist

- [x] My code follows the project's style guidelines
- [x] I have performed a self-review
- [ ] I have added/updated documentation as needed
- [x] My changes generate no new warnings
- [x] Tests pass locally

### CODE REVIEW

**Review by ivantha — COMMENTED** [review]
## ◎ Argus Review

This PR introduces a clean, well-structured registry-driven approach for layer defaults and validation, eliminating hardcoded parameter maps and replacing them with a central `LAYER_REGISTRY`. The Sidebar refactoring is a notable improvement in maintainability. However, there is a high-severity aliasing bug where `layer.defaultParams` is assigned by reference rather than cloned, which will corrupt the registry state during a user session as node params are edited. A medium-severity regression allows numeric `0` to pass required-param validation (the old truthy check caught it; the new strict equality check does not). An unexplained new dependency (`@radix-ui/react-dropdown-menu`) with no visible usage also warrants clarification.

**Suggested Version Bump:** `minor`

### ⚠️ Breaking Changes

- **[INFO]** Behavioral change: numeric 0 in required params no longer blocks save: **Previous behavior**: `canSaveModel` used truthy checks (`!value`) which treated `0` as an invalid/empty value, blocking the save action.

**New behavior**: The registry-driven validation only checks for `undefined`, `null`, and `""`. A numeric `0` in any required param (e.g., `units: 0`) is now considered valid.

**Who is affected**: Any consumer or test that relied on `0` being treated as an invalid required-param value. Users can now set `units = 0` and save the model, which may cause backend/TensorFlow errors at model-build time.

**Migration path**: Add `|| value === 0` to the validation check (see finding on numeric zero regression), or introduce numeric range validation via the registry's param metadata.
- **[LOW]** canSaveModel now silently passes for unregistered node types: **Previous behavior**: Unknown node types in `canSaveModel` fell through to no validation (implicitly valid), but the logic was embedded inline and the set of known types was fixed.

**New behavior**: Any node type not present in `LAYER_REGISTRY` is explicitly skipped via `continue`. This is functionally equivalent today, but the behavior is now entirely controlled by the registry — removing a type from the registry would silently disable its validation.

**Who is affected**: Any code that adds custom or plugin-defined node types to the canvas without registering them.

**Migration path**: Ensure all node types used in the application are registered in `LAYER_REGISTRY`. Consider adding a development-mode warning for unregistered types.


### PR Description Issues

- **[MEDIUM]** PR description shows a different registry structure than what was implemented: The PR description includes a registry example with rich per-param metadata:

```js
params: {
  units: { type: "number", required: true, default: 64, min: 1 },
  activation: { type: "select", required: true, default: "relu", options: ["relu", "sigmoid", "tanh"] }
}
```

However, the actual implementation uses a flat structure with `defaultParams` (an object keyed by param name) and `requiredParams` (an array of strings), with no `type`, `min`, or `options` metadata. The default values are all empty strings (`""`), not meaningful values like `64` or `"relu"`. This mismatch could mislead reviewers into approving a richer feature than what was actually delivered.
  > **Suggestion:** Update the PR description's registry example to reflect the actual implementation:

```js
export const LAYER_REGISTRY = [
  {
    type: "customdense",
    label: "Dense",
    accentClass: "border-l-node-dense",
    defaultParams: { units: "", activation: "" },
    requiredParams: ["units", "activation"],
  },
  // ...
];
```

Or, if the richer schema was the intended design, note in the description that it was descoped and link a follow-up issue.
- **[LOW]** Unchecked documentation checkbox and empty Screenshots section: The PR checklist has `- [ ] I have added/updated documentation as needed` left unchecked with no explanation. Since this PR introduces a new `registry/` directory that is now the authoritative source for layer configuration, some documentation (at minimum a short README or inline JSDoc) would help future contributors understand the contract.

Additionally, the **Screenshots** section is present but completely empty. If the UI changed (the sidebar now renders from the registry), a before/after screenshot would add value.
  > **Suggestion:** Either:
- Check the documentation checkbox and add at least a comment block in `layers.js` explaining the registry contract (how to add a new layer).
- Or explicitly note "No documentation update needed" and explain why.

For Screenshots: add a screenshot of the sidebar/canvas with the new registry-driven rendering, or remove the section if it is not applicable.
